[DataGrid] Dynamic virtualization range

romgrk opened the pull request:
Closes #11344

Implement both improvements proposed in the linked issue. Read the comments below for more implementation details.

Before: https://next.mui.com/x/react-data-grid/#pro-plan
After: https://deploy-preview-12353--material-ui-x.netlify.app/x/react-data-grid/#pro-plan

Before/After:

[Kooha-2024-03-10-20-48-32.webm](https://github.com/mui/mui-x/assets/1423607/0c916eee-a030-4045-976a-f71aa5e7b5a5)

romgrk commented on packages/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx:
```diff
@@ -119,42 +141,54 @@ export const useGridVirtualScroller = () => {
+    scrollCache.lastPosition = scrollPosition.current;
+
+    if (isScrolling) {
+      const dy = scrollPosition.current.top - previousScroll.current.top;
+      const dx = scrollPosition.current.left - previousScroll.current.left;
+      scrollCache.buffer = {
+        top: dimensions.rowHeight * 10,
+        left: 50 * 3,
+      };
```
I tried predicting* the optimal buffer size but failed, so I just put a static 10 rows or 3 columns buffer.

*I had code in place to collect scroll points as event, which I hoped we could run some math on to figure out where the next scroll event would be. I tried linear regression, polynomial regression, and even a small (<1kb) ML model (simple perceptron architecture) with synaptic.js, but the problem is the data. One would expect scroll events to mostly fit to a bell curve. Which they do as a whole. But the intervals and delta values are just too irregular from one point to the next to be able to predict the next point in a sequence. I have gathered some data (me scrolling the grid) and [plotted it here](https://observablehq.com/d/bf8a4012ffdb8453), the last graph is an example of "generally like a bell curve but unpredictable from one point to the next". Maybe with `wheel` events added in the mix we could get a better error rate (best I could get was 50% error on average), but that will take some more research.

MBilalShafi replied:
I can't access the link attached. Could you update the permissions?

romgrk replied:
Done

romgrk commented on packages/x-data-grid/src/models/props/DataGridProps.ts:
```diff
@@ -128,30 +128,30 @@ export interface DataGridPropsWithDefaultValues {
+   * @default 100
    */
-  columnBuffer: number;
+  columnBufferPx: number;
   /**
-   * Number of extra rows to be rendered before/after the visible slice.
-   * @default 3
+   * Row region to render before/after the viewport
+   * @default 100
    */
-  rowBuffer: number;
+  rowBufferPx: number;
```
I don't love the new name with a `Px` suffix, but leaving the name as it didn't feel like a good option because it would silently make the grid less performant for the users using this prop.

MBilalShafi replied:
`rowBufferRegion` could be another option, JSDoc can explain the unit of the region (px)

romgrk commented on packages/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx:
```diff
@@ -37,10 +37,28 @@ import {
 
 export type VirtualScroller = ReturnType<typeof useGridVirtualScroller>;
 
+enum ScrollDirection {
+  NONE,
+  UP,
+  DOWN,
+  LEFT,
+  RIGHT,
+}
+
+const EMPTY_SCROLL_CACHE = {
+  direction: ScrollDirection.NONE,
+  lastTimestamp: -1,
+  lastPosition: { top: 0, left: 0 },
+  buffer: { top: 0, left: 0 },
+};
+type ScrollCache = typeof EMPTY_SCROLL_CACHE;
```
The general idea for the logic is that scroll events trigger right after a new frame has been rendered, so too late for us to prevent white areas that way. If we want to prevent white areas, we need to diplay enough rows/columns so that there is enough of them to cover the area from the current viewport to where the viewport will be at the next scroll event.

As mentioned above predicting by how much the user will have scrolled at the next scroll event is a problem is itself, but by having a substantially large buffer in one direction we try to ensure users see as few white areas as possible.

MBilalShafi commented on packages/x-data-grid/src/DataGrid/DataGrid.tsx:
```diff
@@ -658,9 +658,9 @@ DataGridRaw.propTypes = {
   rowSpacingType: PropTypes.oneOf(['border', 'margin']),
   /**
    * Number of rows from the `rowBuffer` that can be visible before a new slice is rendered.
```
```suggestion
   * Number of pixels from the `rowBufferPx` that can be visible before a new slice is rendered.
```

MBilalShafi commented on docs/data/data-grid/demo/FullFeaturedDemo.js:
Additionally, a row virtualization demo and explaining a bit about `rowThresholdPx` and `columnThresholdPx` would be great too.
```diff
diff --git a/docs/data/data-grid/virtualization/virtualization.md b/docs/data/data-grid/virtualization/virtualization.md
index 7ee018b63..a4527d228 100644
--- a/docs/data/data-grid/virtualization/virtualization.md
+++ b/docs/data/data-grid/virtualization/virtualization.md
@@ -11,7 +11,7 @@ _\*unlimited: Browsers set a limit on the number of pixels a scroll container ca
 
 Row virtualization is the insertion and removal of rows as the data grid scrolls vertically.
 
-The grid renders twice as many rows as are visible. It isn't configurable yet.
+The grid renders some additional rows above and below the visible rows. You can use `rowBufferPx` prop to adjust the area to scan (in pixels). The default value is `150`.
 Row virtualization is limited to 100 rows in the `DataGrid` component.
 
 ## Column virtualization
@@ -23,11 +23,11 @@ Column virtualization is the insertion and removal of columns as the data grid s
 - Overscanning more allows the built-in search feature of the browser to find more matching cells.
 - Overscanning too much can negatively impact performance.
 
-By default, 2 columns are rendered outside of the viewport. You can change this option with the `columnBuffer` prop. The following demo renders 1,000 columns in total:
+By default, columns coming under 150 pixels region are rendered outside of the viewport. You can change this option with the `columnBufferPx` prop. The following demo renders 1,000 columns in total:
 
 {{"demo": "ColumnVirtualizationGrid.js", "bg": "inline"}}
 
-You can disable column virtualization by setting the column buffer to a higher number than the number of rendered columns, e.g. with `columnBuffer={columns.length}` or `columnBuffer={Number.MAX_SAFE_INTEGER}`.
+You can disable column virtualization by setting the column buffer to a higher number than the number of rendered columns, e.g. with `columnBufferPx={Number.MAX_SAFE_INTEGER}`.
 
 ## Disable virtualization
```

MBilalShafi commented on docs/data/migration/migration-data-grid-v6/migration-data-grid-v6.md:
```diff
@@ -106,9 +106,10 @@ As a result, the following changes have been made:
+## Renamed props
 
-- -->
+- The props `rowBuffer`, `columnBuffer`, `rowThreshold` and `columnThreshold` were renamed to `rowBufferPx`, `columnBufferPx`, `rowThresholdPx` and `columnThresholdPx`.
```
Nit: We might also call it the removal of some props and addition of some to make things more explicit. Since the API and expected values are updated.

romgrk replied:
I think the rename entry with the explanation is fine. The 2 removed props have been moved to their section.

MBilalShafi reviewed (approved):
Nice improvement 🚀 
Looks smoother

romgrk commented on packages/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx:
```diff
@@ -76,11 +94,29 @@ export const useGridVirtualScroller = () => {
 
   useResizeObserver(mainRef, () => apiRef.current.resize());
 
+  /*
+   * Scroll context logic
+   * ====================
+   * We only render the cells contained in the `renderContext`. However, when the user starts scrolling the grid
+   * in a direction, we want to render as many cells as possible in that direction, as to avoid presenting white
+   * areas if the user scrolls too fast/far and the viewport ends up in a region we haven't rendered yet. To render
+   * more cells, we store some offsets to add to the viewport in `scrollCache.buffer`. Those offsets make the render
+   * context wider in the direction the user is going, but also makes the buffer around the viewport `0` for the
+   * dimension (horizontal or vertical) in which the user is not scrolling. So if the normal viewport is 8 columns
+   * wide, with a 1 column buffer (10 columns total), then we want it to be exactly 8 columns wide during vertical
+   * scroll.
+   * However, we don't want the rows in the old context to re-render from e.g. 10 columns to 8 columns, because that's
+   * work that's not necessary. Thus we store the context at the start of the scroll in `frozenContext`, and the rows
+   * that are part of this old context will keep their same render context as to avoid re-rendering.
+   */
```
![image](https://github.com/mui/mui-x/assets/1423607/0f75d837-59b2-4ba3-9265-8e820b62635a)

cherniavskii replied:
Thanks for the detailed comment and visualization!

romgrk replied:
Sure, I couldn't find names good enough to explain the logic alone. And I made the graph before implementing the thing, helps my brain understand which objects/concepts I'll need. Also I just like graphs.

MBilalShafi replied:
Yes, pretty useful to understand what's going on 👍

MBilalShafi replied:
> but also makes the buffer around the viewport `0` for the dimension (horizontal or vertical) in which the user is not scrolling

Nice. Would this work on iOS too? It might cause issues due to the [diagonal scrolling](https://github.com/mui/mui-x/issues/11230), but it will probably be solved automatically when we solve 11230, right?

romgrk replied:
Yes, we'll need to solve diagonal scrolling, but that shouldn't be too hard.

romgrk commented:
I have removed the threshold props as discussed on slack.

@MBilalShafi I have updated substantially since your review, you might want to review again.

MBilalShafi commented on packages/x-data-grid/src/models/props/DataGridProps.ts:
```diff
@@ -129,12 +129,12 @@ export interface DataGridPropsWithDefaultValues {
   /**
    * Row region in pixels to render before/after the viewport
-   * @default 100
+   * @default 150
```
Should we initialize it based on the `props.rowHeight`, maybe `props.rowHeight * 3` or something?
I want to [make](https://github.com/mui/mui-x/pull/12317/files#diff-237b220ab18e9e327a1bdd56eba704ca2f7fbb51b48b57c7973025827b17fc15R22-R24) the initial props' object a function in the datasource feature (a potential BC), which could be helpful in deducing initial props based on other props, we might use it to initialize `rowBufferPx`, wdyt?

romgrk replied:
We'd need to set the default value to `undefined`, then we'd be able to pick the best value when the logic runs. Does that sound good?

romgrk replied:
I'll leave it as it is for now, changing the default value later is not a breaking change I think.

MBilalShafi replied:
Sound good 👍

cherniavskii commented on packages/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx:
```diff
@@ -124,62 +178,88 @@ export const useGridVirtualScroller = () => {
+      rootProps.rowBufferPx,
+      rootProps.columnBufferPx,
+      dimensions.rowHeight * 15,
+      50 * 3,
```
Why is the horizontal buffer much smaller than the vertical one?
Assuming the default row height, it's 780px vs only 150px.

romgrk replied:
Could be increased, but those numbers felt good in practice. If you have a different proposition I'm ok with trying it out.

cherniavskii replied:
How about 300px?

cherniavskii commented on packages/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx:
```diff
@@ -557,7 +657,11 @@ function computeRenderContext(inputs: RenderContextInputs, scrollPosition: Scrol
     const firstRowIndex = Math.min(
-      getNearestIndexToRender(inputs, top),
+      getNearestIndexToRender(inputs, top, {
+        atStart: true,
+        lastPosition:
+          inputs.rowsMeta.positions[inputs.rowsMeta.positions.length - 1] + inputs.lastRowHeight,
```
Should we avoid the object wrapper here for performance reasons?
```diff
- options: SearchOptions | undefined = undefined,
+ lastPosition: number | undefined, // Or call it `startPosition`?
```

romgrk replied:
This code is called once per render context update so I'm less worried because it's not multiplied by the number of rows/cells. The param name is ok as `lastPosition` considering what we do with it, but I'll check if I can simplify this logic.

cherniavskii replied:
Feel free to discard then, I assumed it's being called more often 👍🏻

romgrk commented on packages/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx:
```diff
@@ -124,62 +178,88 @@ export const useGridVirtualScroller = () => {
+
+    // Since previous render, we have scrolled...
+    const rowScroll = Math.abs(
+      scrollPosition.current.top - previousContextScrollPosition.current.top,
```
`previousContextScrollPosition` is the scroll position the last time the render context was updated, not just the "previous scroll position". That's why the long name. It's needed for the row/column threshold (of length 1) calculation (which is still there even if we don't expose it as a configurable prop). The current/previous positions are there for the direction logic, which is needed for the virtualization range.

cherniavskii replied:
Alright, I missed the "context" part of the name 👍🏻

romgrk commented on packages/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx:
```diff
@@ -43,20 +45,53 @@ import {
+
+const isJSDOM = typeof window !== 'undefined' ? /jsdom/.test(window.navigator.userAgent) : false;
+
 export const useGridVirtualScroller = () => {
   const apiRef = useGridPrivateApiContext() as React.MutableRefObject<PrivateApiWithInfiniteLoader>;
   const rootProps = useGridRootProps();
   const visibleColumns = useGridSelector(apiRef, gridVisibleColumnDefinitionsSelector);
-  const enabled = useGridSelector(apiRef, gridVirtualizationEnabledSelector);
-  const enabledForColumns = useGridSelector(apiRef, gridVirtualizationColumnEnabledSelector);
+  const enabled = useGridSelector(apiRef, gridVirtualizationEnabledSelector) && !isJSDOM;
+  const enabledForColumns =
+    useGridSelector(apiRef, gridVirtualizationColumnEnabledSelector) && !isJSDOM;
```
This PR broke about 30 tests in jsdom. It was this or adding `disableVirtualization={isJSDOM}` in those 30 tests.

cherniavskii replied:
> This PR broke about 30 tests in jsdom.

Do you have an idea why?

romgrk replied:
No, but seeing how this single change fixed them all it's virtualization for sure.
